### PR TITLE
AMDGPU: Fix inst-selection of large scratch offsets with sgpr base

### PR DESCRIPTION
**Pull request by petar-avramovic**
Use i32 for offset instead of i16, this way it does not get interpreted
as negative 16 bit offset.

### CODE REVIEW

File: llvm/lib/Target/AMDGPU/AMDGPUISelDAGToDAG.cpp [hunk 1]
```diff
@@ -1911,7 +1911,7 @@ bool AMDGPUDAGToDAGISel::SelectScratchSAddr(SDNode *Parent, SDValue Addr,
                     0);
   }
 
-  Offset = CurDAG->getTargetConstant(COffsetVal, DL, MVT::i16);
+  Offset = CurDAG->getTargetConstant(COffsetVal, DL, MVT::i32);
```
Comment (jayfoad) [inline llvm/lib/Target/AMDGPU/AMDGPUISelDAGToDAG.cpp:1914]: There are two more places in `SelectScratchSVAddr` below that need the same fix.

Comment (petar-avramovic) [reply llvm/lib/Target/AMDGPU/AMDGPUISelDAGToDAG.cpp:1914]: fixed and added tests for those cases

**Review by jayfoad — APPROVED** [review]
LGTM, thanks! Please also backport to release/19.x.

**Review by arsenm — APPROVED** [review]

**Comment by petar-avramovic** [comment]
### Merge activity

* **Sep 30, 4:43 AM EDT**: @petar-avramovic started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/llvm/llvm-project/110256).

**Pull request merged**